samples/time/timer.cpp

```diff
@@ -0,0 +1,75 @@
+#include <iostream>
+#include <string>
+
+#include <Wink/address.h>
+#include <Wink/machine.h>
+#include <Wink/state.h>
+
+int main(int argc, char **argv) {
+  if (argc < 3) {
+    error() << "Incorrect parameters, expected <spawner> <address> <duration>\n"
+            << std::flush;
+    return -1;
+  }
+
+  Address spawner(argv[1]);
+  Address address(argv[2]);
+  UDPSocket socket;
+  Machine m(spawner, address, "time/Timer", socket);
+
+  std::istringstream iss(argv[3]);
+  int seconds;
+  iss >> seconds;
+  std::time_t start = std::time(nullptr);
+
+  m.AddState(std::make_unique<State>(
+      // State Name
+      "idle",
+      // Parent State
+      "",
+      // On Entry Action
+      []() { info() << "Timer is IDLE\n"
+                    << std::flush; },
+      // On Exit Action
+      []() {},
+      // Receivers
+      std::map<const std::string, Receiver>{
+          {"idle", [&](const Address &sender,
+                       std::istream &args) { m.GotoState("idle"); }},
+          {"start",
+           [&](const Address &sender, std::istream &args) {
+             start = std::time(nullptr);
+             m.GotoState("timing");
+           }},
+          {"stop",
+           [&](const Address &sender, std::istream &args) { m.Exit(); }},
+          {"exit",
+           [&](const Address &sender, std::istream &args) { m.Exit(); }},
+      }));
+
+  m.AddState(std::make_unique<State>(
+      // State Name
+      "timing",
+      // Parent State
+      "",
+      // On Entry Action
+      [&]() {
+        const auto now = std::time(nullptr);
+        const auto elapsed = now - start;
+        info() << "Timer is TIMING: "<<elapsed<<'\n' << std::flush;
+        if (elapsed < seconds) {
+          sleep(1);
```

Comment:
This will cause the SM to not be lively for 1 second, i.e. it will not process any messages for the duration of the sleep. As a sample this is fine, but I'd suggest adding a comment that in a production SM they should never sleep, perform long computations or do blocking IO. These types of operations need to be delegated to other threads/background processes which would send messages to the SM to signal completion.

Eventually, there should be an example solving this type of problem and in a "Guide/Book" this type of information should be well documented.